demo/src/ddnet/writer.rs

```diff
@@ -0,0 +1,205 @@
+    last_tick: i32,
+    /// For delta timing
+    last_keyframe: Option<i32>,
+    uuid_index: UuidIndex,
+    snap: snapshot::Snap,
+    builder: snap::Builder,
+    delta: snapshot::Delta,
+    buf: arrayvec::ArrayVec<[u8; format::MAX_SNAPSHOT_SIZE]>,
+    i32_buf: Vec<i32>,
+}
+
+struct UuidIndex {
+    index: HashMap<[u8; 16], u16>,
+    next_type_id: u16,
+    advertised: HashSet<[u8; 16]>,
+}
+
+impl UuidIndex {
+    fn new() -> Self {
+        Self {
+            index: HashMap::new(),
+            next_type_id: 32767,
+            advertised: HashSet::new(),
+        }
+    }
+
+    /// The uuid items need to be inserted into every snap.
+    /// Not only in the first one where they appear.
+    /// This function should be called inbetween each snap.
+    fn reset(&mut self) {
+        self.advertised.clear();
+    }
+
+    fn get_type_id(
+        &mut self,
+        builder: &mut snap::Builder,
+        uuid: &[u8; 16],
+    ) -> Result<u16, snap::BuilderError> {
+        let type_id = match self.index.entry(*uuid) {
+            hash_map::Entry::Occupied(type_id) => *type_id.get(),
+            hash_map::Entry::Vacant(uuid_entry) => {
+                let type_id = self.next_type_id;
+                uuid_entry.insert(type_id);
+                self.next_type_id -= 1;
+                type_id
+            }
+        };
+        if !self.advertised.contains(uuid) {
+            self.advertised.insert(*uuid);
+            let mut uuid_item_data = [0; 4];
+            for (mut_data, uuid_data) in uuid_item_data.iter_mut().zip(uuid.chunks(4)) {
+                *mut_data = i32::from_be_bytes(uuid_data.try_into().unwrap());
+            }
+            builder.add_item(0, type_id, &uuid_item_data)?;
+        }
+        Ok(type_id)
+    }
+}
+
+fn mem_swap_default<T: Default>(mut_ref: &mut T) -> T {
+    let mut swap = T::default();
+    mem::swap(&mut swap, mut_ref);
+    swap
+}
+
+impl DemoWriter {
+    pub fn new<T: io::Write + io::Seek + 'static>(
+        file: T,
+        net_version: &[u8],
+        map_name: &[u8],
+        map_sha256: Option<common::digest::Sha256>,
+        map_crc: u32,
+        kind: crate::DemoKind,
+        length: i32,
+        timestamp: &[u8],
+        map: &[u8],
+    ) -> Result<Self, WriteError> {
+        let raw = crate::Writer::new(
+            file,
+            net_version,
+            map_name,
+            map_sha256,
+            map_crc,
+            kind,
+            length,
+            timestamp,
+            map,
+        )?;
+
+        Ok(Self {
+            inner: raw,
+            last_tick: -1,
+            last_keyframe: None,
+            uuid_index: UuidIndex::new(),
+            snap: snapshot::Snap::default(),
+            delta: snapshot::Delta::default(),
+            builder: snap::Builder::default(),
+            buf: arrayvec::ArrayVec::new(),
+            i32_buf: Vec::new(),
+        })
+    }
+
+    pub fn write_snap<'a, T: Iterator<Item = &'a (snap_obj::SnapObj, u16)>>(
```

Comment:
```suggestion
    pub fn write_snap<'a, T: Iterator<Item = (&'a snap_obj::SnapObj, u16)>>(
```
would be more general.